### ISSUE TITLE
Provide mechanism for changing streaming delimiter from standard API

### ISSUE THREAD

**Issue by srithon** [cohere-ai/cohere-go#72]
Currently, it seems that you can change the delimiter used for buffered reads from streaming responses when you directly call `NewStream(response, opts...)` to instantiate streams; however, just using the `ChatStream` method on `Client`, there doesn't seem to be a way to set the delimiter. In `ChatStream`, to add a parameter for `delimiter` you would just need to pass a `delimiter` option into the instantiation of `&core.StreamParams` at the end of the method, when creating/returning the `Stream`. I think it would be nice to have this option without having to copy all of the code in `ChatStream`. I checked the Python SDK and it seems that it has the same limitation with the standard API, so this change may be wider in scope than I thought. With that said, would it be possible/desirable to make the aforementioned change to add `delimiter` to the "public" public API? (the part that most people will be using)

**Comment by billytrend-cohere** [comment]
hey @srithon, thanks for reporting. Trying to understand the issue. What are you trying to achieve by changing the delimeter? I assume you mean the \n between streamed events. We are currently looking at adopting SSE, would that solve your issue?

**Comment by srithon** [comment]
> hey @srithon, thanks for reporting. Trying to understand the issue. What are you trying to achieve by changing the delimeter? I assume you mean the \n between streamed events. We are currently looking at adopting SSE, would that solve your issue? 

Yea, sorry, I should've been more clear. I was indeed talking about the \n between events. With the delimiter set to the newline character, the smallest unit of text the client can process at once is an entire line/paragraph. For my UI, I wanted a chatbot-like interface where responses are streamed in word-by-word rather than line-by-line. While this **effect** can be emulated by the frontend (just splitting the received line on white space and displaying the words with a randomized delay between them), the main difference is increased _perceived_ TTFT (time to first token) by the user; if the buffered reader's delimiter could be set to a space character instead of a newline, the client could _display_ the first output token/word well before the entire first line was returned by the API, making the end application appear to have lower latency!

With regards to SSE, I'm afraid I would need a bit more detail to say whether it addresses my issue; if by "adopting SSE" you mean that the client, when using the ChatStream endpoint, would open an SSE connection with Cohere's server (via EventSource or a thin wrapper), specifying the generation id for the given request, and would then receive messages from the server as they were generated without having the Cohere client-side SDK buffering the responses, then I believe it could resolve the issue. The main issue, to reiterate, is that the current primary/documented method for streaming chat responses does not allow the client to _process_ units of text smaller than discrete lines/paragraphs; while the server-side does appear to send generated text at a higher granularity, the invariable line-by-line buffering of the current client SDK implementation prevents developers from taking advantage of this granularity.

Please let me know if you have any further questions.

Thanks,
Sridaran

**Comment by srithon** [comment]
Hey @billytrend-cohere, sorry for the late reply, but I just realized that this issue was completely nonsensical and off-base; I apologize for the misunderstanding. What happened is that I had written a test app using Cohere's Python SDK, where the documentation had this code:
```python
for event in stream:
    if event.event_type == "text-generation":
        print(event.text, end='')
```
When running this code, I noticed that text was only outputting line-by-line rather than word-by-word, and so I inferred that text generation was line-by-line. However, in reality what was happening is that Python was buffering the output and only flushing the buffer when receiving a newline character, causing text to be displayed to the terminal line-by-line rather than per-generation. I filed a [PR](https://github.com/cohere-ai/cohere-python/pull/487) under the Python SDK repository to explicitly add output flushing to the streaming example, to prevent anyone else from having this confusion.

I filed the issue under this repo (the Golang SDK) because I was working on a hackathon project in Golang, and figured that since both SDKs had equivalent APIs, that the Golang one would have the same issue. After digging through the code a bit, I stumbled upon the red herring of the `delimiter`  field and interpreted it incorrectly, as you realized.

Thanks for your time in helping get to the bottom of the issue! Other than this mixup, I had a great experience using the Cohere SDK, and commend your dedication to the community.
